### DISCUSSION TITLE
Mypy: AsyncSession "is not valid as a type"

### DISCUSSION THREAD

**Discussion by gwerbin** [sqlalchemy/sqlalchemy#10113] [Usage Questions] [answered]
I am using CPython 3.10.11, SQLAlchemy 2.0.19, and Mypy 1.4.1.

Here is the Mypy section of my `pyproject.toml` file:

```toml
[tool.mypy]
mypy_path = ["source"]
python_version = "3.10"
strict = true

plugins =  [
  "numpy.typing.mypy_plugin",
  "pydantic.mypy",
]
```

I am creating an `async_sessionmaker`  as follows:

```python
from datetime import datetime as Datetime
from typing import AsyncGenerator

from sqlalchemy import TIMESTAMP
from sqlalchemy.ext.asyncio import AsyncEngine
from sqlalchemy.ext.asyncio import AsyncSession as _AsyncSession
from sqlalchemy.ext.asyncio import async_sessionmaker, create_async_engine
from sqlalchemy.orm import DeclarativeBase


class Base(DeclarativeBase):
    type_annotation_map = {
        Datetime: TIMESTAMP(timezone=True),
    }


async_engine = create_async_engine(
    "...",
    pool_pre_ping=True,
    pool_size=20,
    pool_recycle=5,
    max_overflow=5,
)

AsyncSession = async_sessionmaker(async_engine)

async def get_db() -> AsyncGenerator[AsyncSession, None]:
    r"""Get a fresh database session for the current request, for use with FastAPI Depends()."""
    async with AsyncSession() as session:
        yield session
```

But I get the following output:

```none
argh.py:26: note: Revealed type is "sqlalchemy.ext.asyncio.session.async_sessionmaker[sqlalchemy.ext.asyncio.session.AsyncSession]"
argh.py:28: error: Variable "argh.AsyncSession" is not valid as a type  [valid-type]
argh.py:28: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
Found 1 error in 1 file (checked 1 source file)
```

This means I need to use `# type:ignore[valid-type]` whenever I want to use this `AsyncSession` in a type hint, which is annoying and reduces the type-safety of my application.

I must be doing something wrong, because I can't find anything on the issue tracker, but this seems like such a basic problem.  Any ideas?

**Comment by zzzeek** [comment, marked as answer]
mypy disallows a variable returned from a function from being used as a type.

Your program is assigning the return value of `async_sessionmaker` to the variable named `AsyncSession`.

this means that `AsyncSession` is now *not* a type.  can't be used in the signature for `get_db()` on the next line.   based on the imports you have here, the correct return type is `AsyncGenerator[_AsyncSession, None]`

```py
async def get_db() -> AsyncGenerator[_AsyncSession, None]:
    r"""Get a fresh database session for the current request, for use with FastAPI Depends()."""
    async with AsyncSession() as session:
        yield session
```

**Reply by gwerbin** [reply]
I didn't realize this was a global Mypy limitation. Using `_AsyncSession` is fine as a workaround, it's not _our_ particular `AsyncSession` type but it's good enough to fix the problem.

**Reply by gwerbin** [reply]
@zzzeek Is this documented behavior in Mypy? similar issues have been treated as bugs in the past, e.g. https://github.com/python/mypy/issues/5865

**Reply by CaselIT** [reply]
> Using _AsyncSession is fine as a workaround, it's not our particular AsyncSession type but it's good enough to fix the problem.

At runtime the type *is* _AsyncSession, so that's not a workaround, it's the correct typing

**Reply by gwerbin** [reply]
> > Using _AsyncSession is fine as a workaround, it's not our particular AsyncSession type but it's good enough to fix the problem.
> 
> At runtime the type _is_ _AsyncSession, so that's not a workaround, it's the correct typing

I see, that clarifies things. I was under the impression that my "AsyncSession" was in fact a new class, based on the conventional styling of capitalizing the session factory, and the fact that there was a distinct revealed type of `sqlalchemy.ext.asyncio.session.async_sessionmaker[sqlalchemy.ext.asyncio.session.AsyncSession]`.

Perhaps a less-confusing name style would be something like this?

```python
from sqlalchemy.ext.asyncio import AsyncSession, async_sessionmaker

make_new_session = async_sessionmaker(async_engine)

async def get_db() -> AsyncGenerator[AsyncSession, None]:
    r"""Get a fresh database session for the current request, for use with FastAPI Depends()."""
    async with make_new_session() as session:
        yield session
```

**Reply by CaselIT** [reply]
That's a variable name, so you can call it however you want.

**Reply by zzzeek** [reply]
the sessionmaker does return a Python class.  it's a class.   pep-484 does not have any support for dynamically generated classes though, so the returned class from such a function needs to be typed either in terms of the class' static superclass (`_AsyncSession` in this case) or if it had lots of special new methods on it, using a `typing.Protocol`.

**Reply by CaselIT** [reply]
> the sessionmaker does return a Python class.

sessionmaker returns an instance of sessionmaker, no? https://github.com/sqlalchemy/sqlalchemy/blob/265667a27bba2247143779ff3f3d141b0e3fface/lib/sqlalchemy/orm/session.py#L4735
it's not a subclass of type
